File: src/lib/goals.ts
// Objectifs d'entraînement (100 % local) : « 3 sorties par semaine »,
// « 100 km par mois », « 5000 kg de tonnage par mois »… Les définitions sont
// stockées en JSON dans la table `settings` (clé `goals`) — donc incluses dans
// la sauvegarde S3 et l'export coach sans migration de schéma — et la
// progression est calculée À LA VOLÉE depuis les séances terminées (mêmes
// filtres que les statistiques existantes : `endedAt IS NOT NULL`).
//
// La logique pure (périodes, progression, libellés, parsing, opérations de
// liste) est testée dans __tests__/lib/goals.test.ts ; l'accès base est isolé
// en fin de module.
import { getSetting, setSetting, statsBetween, tonnageBetween } from '@/lib/db';
import { nowMs } from '@/lib/time';
import { startOfWeekMs } from '@/lib/week';

/** Métrique suivie : nombre de séances, distance vélo (km) ou tonnage muscu (kg). */
export type GoalMetric = 'sessions' | 'distance' | 'tonnage';
/** Fenêtre de suivi : semaine (lundi→dimanche) ou mois calendaire. */
export type GoalPeriod = 'week' | 'month';
/** Type d'activité compté (pertinent pour la métrique `sessions` uniquement). */
export type GoalActivity = 'all' | 'velo' | 'muscu';

export type Goal = {
  id: string;
  metric: GoalMetric;
  period: GoalPeriod;
  /** Cible dans l'unité naturelle : séances (n), km (distance), kg (tonnage). */
  target: number;
  /** Restreint le comptage `sessions` à un type ; ignoré pour distance/tonnage. */
  activity: GoalActivity;
};

const SETTING_KEY = 'goals';
const WEEK_MS = 7 * 86_400_000;

const METRICS: GoalMetric[] = ['sessions', 'distance', 'tonnage'];
const PERIODS: GoalPeriod[] = ['week', 'month'];
const ACTIVITIES: GoalActivity[] = ['all', 'velo', 'muscu'];

// ---------------------------------------------------------------------------
// Logique pure (périodes, progression, libellés, parsing, opérations de liste)
// ---------------------------------------------------------------------------

/** Bornes `[fromMs, toMs)` de la période courante contenant `now` (ms epoch). */
export function periodRange(period: GoalPeriod, now: number): { fromMs: number; toMs: number } {
  if (period === 'week') {
    const fromMs = startOfWeekMs(now);
    return { fromMs, toMs: fromMs + WEEK_MS };
  }
  // Mois calendaire local : du 1er du mois au 1er du mois suivant (le
  // dépassement d'index de mois est géré nativement par Date).
  const d = new Date(now);
  const fromMs = new Date(d.getFullYear(), d.getMonth(), 1).getTime();
  const toMs = new Date(d.getFullYear(), d.getMonth() + 1, 1).getTime();
  return { fromMs, toMs };
}

export type GoalProgress = {
  goal: Goal;
  /** Valeur réalisée sur la période, dans l'unité de la cible. */
  value: number;
  target: number;
  /** Avancement borné à [0, 1] (pour une barre/anneau). */
  ratio: number;
  /** Objectif atteint ou dépassé. */
  done: boolean;
};

/** Construit l'avancement d'un objectif à partir de la valeur réalisée. */
export function computeProgress(goal: Goal, value: number): GoalProgress {
  const target = goal.target;
  const ratio = target > 0 ? Math.min(1, Math.max(0, value / target)) : 0;
  return { goal, value, target, ratio, done: target > 0 && value >= target };
}

const PERIOD_LABEL: Record<GoalPeriod, string> = { week: 'semaine', month: 'mois' };
const ACTIVITY_NOUN: Record<GoalActivity, string> = {
  all: 'séances',
  velo: 'sorties vélo',
  muscu: 'séances muscu',
};

/** Libellé court d'un objectif : « 3 sorties vélo / semaine », « 100 km / mois ». */
export function describeGoal(goal: Goal): string {
  const per = PERIOD_LABEL[goal.period];
  if (goal.metric === 'distance') return `${goal.target} km / ${per}`;
  if (goal.metric === 'tonnage') return `${goal.target} kg soulevés / ${per}`;
  return `${goal.target} ${ACTIVITY_NOUN[goal.activity]} / ${per}`;
}

/** Valeur réalisée formatée avec son unité (« 2 », « 42,5 km », « 5000 kg »). */
export function formatGoalValue(goal: Goal, value: number): string {
  if (goal.metric === 'distance') return `${value} km`;
  if (goal.metric === 'tonnage') return `${Math.round(value)} kg`;
  return String(value);
}

/** Valide et normalise une entrée inconnue en `Goal` (ou `null` si invalide). */
function normalizeGoal(input: unknown): Goal | null {
  if (!input || typeof input !== 'object') return null;
  const o = input as Record<string, unknown>;
  if (!METRICS.includes(o.metric as GoalMetric)) return null;
  if (!PERIODS.includes(o.period as GoalPeriod)) return null;
  const target = Number(o.target);
  if (!Number.isFinite(target) || target <= 0) return null;
  const activity = ACTIVITIES.includes(o.activity as GoalActivity)
    ? (o.activity as GoalActivity)
    : 'all';
  const id = typeof o.id === 'string' && o.id.length > 0 ? o.id : makeId();
  return { id, metric: o.metric as GoalMetric, period: o.period as GoalPeriod, target, activity };
}

/** Lit une liste d'objectifs depuis la valeur de réglage (robuste : ignore le
 * JSON corrompu ou les entrées invalides plutôt que de planter). */
export function parseGoals(raw: string | null): Goal[] {
  if (!raw) return [];
  let arr: unknown;
  try {
    arr = JSON.parse(raw);
  } catch {
    return [];
  }
  if (!Array.isArray(arr)) return [];
  const out: Goal[] = [];
  for (const item of arr) {
    const goal = normalizeGoal(item);
    if (goal) out.push(goal);
  }
  return out;
}

export function serializeGoals(goals: Goal[]): string {
  return JSON.stringify(goals);
}

/** Ajoute un objectif, ou remplace celui de même `id`. */
export function upsertGoal(goals: Goal[], goal: Goal): Goal[] {
  const i = goals.findIndex((g) => g.id === goal.id);
  if (i === -1) return [...goals, goal];
  const next = goals.slice();
  next[i] = goal;
  return next;
}

export function removeGoal(goals: Goal[], id: string): Goal[] {
  return goals.filter((g) => g.id !== id);
}

/** Crée un objectif avec un identifiant frais. */
export function makeGoal(input: Omit<Goal, 'id'>): Goal {
  return { id: makeId(), ...input };
}

function makeId(): string {
  return `g${Date.now().toString(36)}${Math.random().toString(36).slice(2, 8)}`;
}

// ---------------------------------------------------------------------------
// Accès base : stockage des définitions + mesure de la progression
// ---------------------------------------------------------------------------

export async function loadGoals(): Promise<Goal[]> {
  return parseGoals(await getSetting(SETTING_KEY));
}

export async function saveGoals(goals: Goal[]): Promise<void> {
  await setSetting(SETTING_KEY, serializeGoals(goals));
}

/** Valeur réalisée d'un objectif sur sa période courante, dans l'unité cible. */
export async function measureGoal(goal: Goal, now: number = nowMs()): Promise<number> {
  const { fromMs, toMs } = periodRange(goal.period, now);
  if (goal.metric === 'sessions') {
    const type = goal.activity === 'all' ? undefined : goal.activity;
    const s = await statsBetween(fromMs, toMs, type);
    return s.sessionCount;
  }
  if (goal.metric === 'distance') {
    const s = await statsBetween(fromMs, toMs, 'velo');
    // km arrondis au dixième, cohérent avec l'affichage des distances.
    return Math.round((s.totalDistanceM / 1000) * 10) / 10;
  }
  return tonnageBetween(fromMs, toMs);
}

/** Avancement de tous les objectifs définis, pour la période courante. */
export async function goalProgressList(now: number = nowMs()): Promise<GoalProgress[]> {
  const goals = await loadGoals();
  return Promise.all(goals.map(async (g) => computeProgress(g, await measureGoal(g, now))));
}

File: app/src/main/java/ovh/battistella/elan/domain/Goals.kt
// Objectifs d'entraînement (100 % local) : « 3 sorties par semaine »,
// « 100 km par mois », « 5000 kg de tonnage par mois »… Les définitions sont
// stockées en JSON dans la table `settings` (clé `goals`) — donc incluses dans
// la sauvegarde S3 et l'export coach sans migration de schéma — et la
// progression est calculée À LA VOLÉE depuis les séances terminées.
//
// Ce fichier porte la logique pure (périodes, progression, libellés, parsing,
// opérations de liste) ; la mesure en base (`measureGoal`) relève de la couche
// données.
package ovh.battistella.elan.domain

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import kotlin.math.roundToLong
import kotlin.random.Random

/** Métrique suivie : nombre de séances, distance vélo (km) ou tonnage muscu (kg). */
enum class GoalMetric(val key: String) {
    SESSIONS("sessions"),
    DISTANCE("distance"),
    TONNAGE("tonnage");

    companion object {
        fun fromKey(key: String?): GoalMetric? = entries.firstOrNull { it.key == key }
    }
}

/** Fenêtre de suivi : semaine (lundi→dimanche) ou mois calendaire. */
enum class GoalPeriod(val key: String) {
    WEEK("week"),
    MONTH("month");

    companion object {
        fun fromKey(key: String?): GoalPeriod? = entries.firstOrNull { it.key == key }
    }
}

/** Type d'activité compté (pertinent pour `sessions` et `distance`). */
enum class GoalActivity(val key: String) {
    ALL("all"),
    VELO("velo"),
    COURSE("course"),
    MARCHE("marche"),
    MUSCU("muscu");

    companion object {
        fun fromKey(key: String?): GoalActivity? = entries.firstOrNull { it.key == key }
    }
}

data class Goal(
    val id: String,
    val metric: GoalMetric,
    val period: GoalPeriod,
    /** Cible dans l'unité naturelle : séances (n), km (distance), kg (tonnage). */
    val target: Double,
    /**
     * Restreint le comptage à un type d'activité — pour `sessions` comme pour
     * `distance` (« 50 km de course par mois »). Ignoré pour `tonnage`, qui n'existe
     * qu'en musculation. `ALL` = toutes activités confondues.
     */
    val activity: GoalActivity,
)

/** Clé de réglage sous laquelle la couche données persiste la liste d'objectifs. */
const val GOALS_SETTING_KEY = "goals"

private const val DAY_MS = 86_400_000L
private const val WEEK_MS = 7 * DAY_MS

// ---------------------------------------------------------------------------
// Logique pure (périodes, progression, libellés, parsing, opérations de liste)
// ---------------------------------------------------------------------------

/** Bornes `[fromMs, toMs)` d'une période. */
data class PeriodRange(val fromMs: Long, val toMs: Long)

/**
 * Bornes `[fromMs, toMs)` de la période courante contenant `now` (ms epoch,
 * heure locale = fuseau par défaut du système, comme `startOfWeekMs`).
 */
fun periodRange(period: GoalPeriod, now: Long): PeriodRange {
    if (period == GoalPeriod.WEEK) {
        val fromMs = startOfWeekMs(now)
        // Fin = début du lundi SUIVANT (re-floor local), pas `from + 168 h` : aux
        // changements d'heure la semaine dure 167 ou 169 h, et un +168 h fixe
        // décalait la borne d'une heure (séances de fin de dimanche mal comptées).
        val toMs = startOfWeekMs(fromMs + WEEK_MS + 43_200_000L) // +7,5 j → lundi suivant
        return PeriodRange(fromMs, toMs)
    }
    // Mois calendaire local : du 1er du mois au 1er du mois suivant.
    val zone = ZoneId.systemDefault()
    val d = Instant.ofEpochMilli(now).atZone(zone).toLocalDate()
    val first = LocalDate.of(d.year, d.month, 1)
    val fromMs = first.atStartOfDay(zone).toInstant().toEpochMilli()
    val toMs = first.plusMonths(1).atStartOfDay(zone).toInstant().toEpochMilli()
    return PeriodRange(fromMs, toMs)
}

data class GoalProgress(
    val goal: Goal,
    /** Valeur réalisée sur la période, dans l'unité de la cible. */
    val value: Double,
    val target: Double,
    /** Avancement borné à [0, 1] (pour une barre/anneau). */
    val ratio: Double,
    /** Objectif atteint ou dépassé. */
    val done: Boolean,
)

/** Construit l'avancement d'un objectif à partir de la valeur réalisée. */
fun computeProgress(goal: Goal, value: Double): GoalProgress {
    val target = goal.target
    val ratio = if (target > 0) (value / target).coerceIn(0.0, 1.0) else 0.0
    return GoalProgress(goal, value, target, ratio, done = target > 0 && value >= target)
}

private val PERIOD_LABEL: Map<GoalPeriod, String> = mapOf(
    GoalPeriod.WEEK to "semaine",
    GoalPeriod.MONTH to "mois",
)

private val ACTIVITY_NOUN: Map<GoalActivity, String> = mapOf(
    GoalActivity.ALL to "séances",
    GoalActivity.VELO to "sorties vélo",
    GoalActivity.COURSE to "sorties course à pied",
    GoalActivity.MARCHE to "sorties marche",
    GoalActivity.MUSCU to "séances muscu",
)

/** Libellé court d'un objectif : « 3 sorties vélo / semaine », « 100 km / mois ». */
fun describeGoal(goal: Goal): String {
    val per = PERIOD_LABEL.getValue(goal.period)
    val target = goal.target.toJsString()
    return when (goal.metric) {
        GoalMetric.DISTANCE -> "$target km / $per"
        GoalMetric.TONNAGE -> "$target kg soulevés / $per"
        GoalMetric.SESSIONS -> "$target ${ACTIVITY_NOUN.getValue(goal.activity)} / $per"
    }
}

/** Valeur réalisée formatée avec son unité (« 2 », « 42.5 km », « 5000 kg »). */
fun formatGoalValue(goal: Goal, value: Double): String = when (goal.metric) {
    GoalMetric.DISTANCE -> "${value.toJsString()} km"
    GoalMetric.TONNAGE -> "${value.roundToLong()} kg"
    GoalMetric.SESSIONS -> value.toJsString()
}

/** Valide et normalise une entrée JSON inconnue en `Goal` (ou `null` si invalide). */
private fun normalizeGoal(input: Any?): Goal? {
    if (input !is JSONObject) return null
    val metric = GoalMetric.fromKey(input.opt("metric") as? String) ?: return null
    val period = GoalPeriod.fromKey(input.opt("period") as? String) ?: return null
    val target = jsToNumber(input.opt("target"))
    if (target == null || !target.isFinite() || target <= 0) return null
    val activity = GoalActivity.fromKey(input.opt("activity") as? String) ?: GoalActivity.ALL
    val rawId = input.opt("id")
    val id = if (rawId is String && rawId.isNotEmpty()) rawId else makeId()
    return Goal(id, metric, period, target, activity)
}

/** Équivalent de `Number(x)` sur les formes plausibles (nombre, chaîne numérique), `null` sinon. */
private fun jsToNumber(v: Any?): Double? = when (v) {
    is Number -> v.toDouble()
    is String -> v.trim().let { if (it.isEmpty()) 0.0 else it.toDoubleOrNull() }
    is Boolean -> if (v) 1.0 else 0.0
    null, JSONObject.NULL -> null
    else -> null
}

/**
 * Lit une liste d'objectifs depuis la valeur de réglage (robuste : ignore le
 * JSON corrompu ou les entrées invalides plutôt que de planter).
 */
fun parseGoals(raw: String?): List<Goal> {
    if (raw.isNullOrEmpty()) return emptyList()
    val arr = try {
        JSONArray(raw)
    } catch (e: JSONException) {
        return emptyList()
    }
    val out = ArrayList<Goal>(arr.length())
    for (i in 0 until arr.length()) {
        val goal = normalizeGoal(arr.opt(i))
        if (goal != null) out.add(goal)
    }
    return out
}

fun serializeGoals(goals: List<Goal>): String {
    val arr = JSONArray()
    for (g in goals) {
        arr.put(
            JSONObject()
                .put("id", g.id)
                .put("metric", g.metric.key)
                .put("period", g.period.key)
                .put("target", g.target)
                .put("activity", g.activity.key),
        )
    }
    return arr.toString()
}

/** Ajoute un objectif, ou remplace celui de même `id`. */
fun upsertGoal(goals: List<Goal>, goal: Goal): List<Goal> {
    val i = goals.indexOfFirst { it.id == goal.id }
    if (i == -1) return goals + goal
    val next = goals.toMutableList()
    next[i] = goal
    return next
}

fun removeGoal(goals: List<Goal>, id: String): List<Goal> = goals.filter { it.id != id }

/** Crée un objectif avec un identifiant frais. */
fun makeGoal(metric: GoalMetric, period: GoalPeriod, target: Double, activity: GoalActivity): Goal =
    Goal(makeId(), metric, period, target, activity)

private const val BASE36 = "0123456789abcdefghijklmnopqrstuvwxyz"

/** `g` + horodatage en base 36 + 6 caractères aléatoires en base 36. */
fun makeId(): String {
    val rand = CharArray(6) { BASE36[Random.nextInt(BASE36.length)] }
    return "g${System.currentTimeMillis().toString(36)}${String(rand)}"
}
